File: XDM_NEON/src/xdman/util/MacUtils.java
package xdman.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.ArrayList;

public class MacUtils {
	public static void open(final File f) throws FileNotFoundException {
		if (!f.exists()) {
			throw new FileNotFoundException();
		}
		try {
			ProcessBuilder pb = new ProcessBuilder();
			pb.command("open", f.getAbsolutePath());
			if (pb.start().waitFor() != 0) {
				throw new FileNotFoundException();
			}
		} catch (Exception e) {
			Logger.log(e);
		}
	}

	public static void openFolder(String folder, String file) throws FileNotFoundException {
		if (file == null) {
			openFolder2(folder);
			return;
		}
		File f = new File(folder, file);
		if (!f.exists()) {
			throw new FileNotFoundException();
		}
		try {
			ProcessBuilder pb = new ProcessBuilder();
			Logger.log("Opening folder: " + f.getAbsolutePath());
			pb.command("open", "-R", f.getAbsolutePath());
			if (pb.start().waitFor() != 0) {
				throw new FileNotFoundException();
			}
		} catch (Exception e) {
			Logger.log(e);
		}
	}

	private static void openFolder2(String folder) {
		try {
			ProcessBuilder builder = new ProcessBuilder();
			ArrayList<String> lst = new ArrayList<String>();
			lst.add("open");
			lst.add(folder);
			builder.command(lst);
			builder.start();
		} catch (Exception e) {
			Logger.log(e);
		}
	}

	public static boolean launchApp(String app, String args) {
		try {
			ProcessBuilder pb = new ProcessBuilder();
			pb.command("open", "-n", "-a", app, "--args", args);
			if (pb.start().waitFor() != 0) {
				throw new FileNotFoundException();
			}
			// Runtime.getRuntime().exec(new String[] { "open \"" + app + "\" " + args });
			return true;
		} catch (Exception e) {
			Logger.log(e);
			return false;
		}
	}

	public static void keepAwakePing() {
		try {
			Runtime.getRuntime().exec("caffeinate -i -t 3");
		} catch (Exception e) {
			Logger.log(e);
		}
	}

	public static void addToStartup() {
		File dir = new File(System.getProperty("user.home"), "Library/LaunchAgents");
		dir.mkdirs();
		File f = new File(dir, "org.sdg.xdman.plist");
		FileOutputStream fs = null;
		try {
			fs = new FileOutputStream(f);
			fs.write(getStartupPlist().getBytes());
		} catch (Exception e) {
			Logger.log(e);
		} finally {
			try {
				if (fs != null)
					fs.close();
			} catch (Exception e2) {
			}
		}
		f.setExecutable(true);
	}

	public static boolean isAlreadyAutoStart() {
		File f = new File(System.getProperty("user.home"), "Library/LaunchAgents/org.sdg.xdman.plist");
		if (!f.exists())
			return false;
		FileInputStream in = null;
		byte[] buf = new byte[(int) f.length()];
		try {
			in = new FileInputStream(f);
			if (in.read(buf) != f.length()) {
				return false;
			}
		} catch (Exception e) {
			Logger.log(e);
		} finally {
			try {
				if (in != null)
					in.close();
			} catch (Exception e2) {
			}
		}
		String str = new String(buf);
		String s1 = getProperPath(System.getProperty("java.home"));
		String s2 = XDMUtils.getJarFile().getAbsolutePath();
		return str.contains(s1) && str.contains(s2);
	}

	public static void removeFromStartup() {
		File f = new File(System.getProperty("user.home"), "Library/LaunchAgents/org.sdg.xdman.plist");
		f.delete();
	}

	public static String getStartupPlist() {
		String str = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n"
				+ "<!DOCTYPE plist PUBLIC \"-//Apple Computer//DTD PLIST 1.0//EN\"\r\n"
				+ "\"http://www.apple.com/DTDs/PropertyList-1.0.dtd\" >\r\n" + "<plist version=\"1.0\">\r\n"
				+ "	<dict>\r\n" + "		<key>Label</key>\r\n" + "		<string>org.sdg.xdman</string>\r\n"
				+ "		<key>ProgramArguments</key>\r\n" + "		<array>\r\n"
				+ "			<string>%sbin/java</string>\r\n" + "			<string>-Xmx1024m</string>\r\n"
				+ "			<string>-Xdock:name=XDM</string>\r\n" + "			<string>-jar</string>\r\n"
				+ "			<!-- MODIFY THIS TO POINT TO YOUR EXECUTABLE JAR FILE -->\r\n"
				+ "			<string>%s</string>\r\n" + "			<string>-m</string>\r\n" + "		</array>\r\n"
				+ "		<key>OnDemand</key>\r\n" + "		<true />\r\n" + "		<key>RunAtLoad</key>\r\n"
				+ "		<true />\r\n" + "		<key>KeepAlive</key>\r\n" + "		<false />\r\n" + "	</dict>\r\n"
				+ "</plist>";
		String s1 = getProperPath(System.getProperty("java.home"));
		String s2 = XDMUtils.getJarFile().getAbsolutePath();
		return String.format(str, s1, s2);
	}

	private static String getProperPath(String path) {
		if (path.endsWith("/"))
			return path;
		return path + "/";
	}

	public static void browseURL(final String url) {
		try {
			ProcessBuilder pb = new ProcessBuilder();
			pb.command("open", url);
			pb.start();// .waitFor();
		} catch (Exception e) {
			Logger.log(e);
		}
	}

	public static void initShutdown() {
		try {
			ProcessBuilder builder = new ProcessBuilder();
			ArrayList<String> lst = new ArrayList<String>();
			lst.add("osascript");
			lst.add("-e");
			lst.add("tell app \"System Events\" to shut down");
			builder.command(lst);
			builder.start();
		} catch (Exception e) {
			Logger.log(e);
		}
	}
}

File: xdm-app/src/main/java/xdm/app/utils/MacUtils.kt
package xdm.app.utils

import xdm.core.util.Logger
import java.io.File

object MacUtils {
    fun open(f: File): Boolean {
        return runCatching {
            if (!f.exists()) {
                return false
            }
            val pb = ProcessBuilder()
            pb.command("open", f.absolutePath)
            if (pb.start().waitFor() != 0) {
                Logger.error("Open failed")
                return false
            }
        }.isSuccess
    }

    fun openFolder(folder: String, file: String?): Boolean {
        return runCatching {
            if (file == null) {
                openFolder2(folder)
                return true
            }
            val f = File(folder, file)
            if (!f.exists()) {
                return false
            }
            val pb = ProcessBuilder()
            Logger.info("Opening folder: " + f.absolutePath)
            pb.command("open", "-R", f.absolutePath)
            if (pb.start().waitFor() != 0) {
                Logger.error("Open folder failed")
            }
        }.isSuccess
    }

    private fun openFolder2(folder: String) {
        val builder = ProcessBuilder()
        val lst = ArrayList<String>()
        lst.add("open")
        lst.add(folder)
        builder.command(lst)
        builder.start()
    }

//    fun launchApp(app: String?, args: String?): Boolean {
//        try {
//            val pb = ProcessBuilder()
//            pb.command("open", "-n", "-a", app, "--args", args)
//            if (pb.start().waitFor() != 0) {
//                throw FileNotFoundException()
//            }
//            // Runtime.getRuntime().exec(new String[] { "open \"" + app + "\" " + args });
//            return true
//        } catch (e: Exception) {
//            Logger.log(e)
//            return false
//        }
//    }
//
//    fun keepAwakePing() {
//        try {
//            Runtime.getRuntime().exec("caffeinate -i -t 3")
//        } catch (e: Exception) {
//            Logger.log(e)
//        }
//    }
//
//    fun addToStartup() {
//        val dir = File(System.getProperty("user.home"), "Library/LaunchAgents")
//        dir.mkdirs()
//        val f = File(dir, "org.sdg.xdman.plist")
//        var fs: FileOutputStream? = null
//        try {
//            fs = FileOutputStream(f)
//            fs.write(startupPlist.toByteArray())
//        } catch (e: Exception) {
//            Logger.log(e)
//        } finally {
//            try {
//                fs?.close()
//            } catch (e2: Exception) {
//            }
//        }
//        f.setExecutable(true)
//    }
//
//    val isAlreadyAutoStart: Boolean
//        get() {
//            val f = File(System.getProperty("user.home"), "Library/LaunchAgents/org.sdg.xdman.plist")
//            if (!f.exists()) return false
//            var `in`: FileInputStream? = null
//            val buf = ByteArray(f.length().toInt())
//            try {
//                `in` = FileInputStream(f)
//                if (`in`.read(buf).toLong() != f.length()) {
//                    return false
//                }
//            } catch (e: Exception) {
//                Logger.log(e)
//            } finally {
//                try {
//                    `in`?.close()
//                } catch (e2: Exception) {
//                }
//            }
//            val str = String(buf)
//            val s1 = getProperPath(System.getProperty("java.home"))
//            val s2 = XDMUtils.getJarFile().absolutePath
//            return str.contains(s1) && str.contains(s2)
//        }
//
//    fun removeFromStartup() {
//        val f = File(System.getProperty("user.home"), "Library/LaunchAgents/org.sdg.xdman.plist")
//        f.delete()
//    }
//
//    val startupPlist: String
//        get() {
//            val str = ("""<?xml version="1.0" encoding="UTF-8"?>
//<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN"
//"http://www.apple.com/DTDs/PropertyList-1.0.dtd" >
//<plist version="1.0">
//	<dict>
//		<key>Label</key>
//		<string>org.sdg.xdman</string>
//		<key>ProgramArguments</key>
//		<array>
//			<string>%sbin/java</string>
//			<string>-Xmx1024m</string>
//			<string>-Xdock:name=XDM</string>
//			<string>-jar</string>
//			<!-- MODIFY THIS TO POINT TO YOUR EXECUTABLE JAR FILE -->
//			<string>%s</string>
//			<string>-m</string>
//		</array>
//		<key>OnDemand</key>
//		<true />
//		<key>RunAtLoad</key>
//		<true />
//		<key>KeepAlive</key>
//		<false />
//	</dict>
//</plist>""")
//            val s1 = getProperPath(System.getProperty("java.home"))
//            val s2 = XDMUtils.getJarFile().absolutePath
//            return String.format(str, s1, s2)
//        }
//
//    private fun getProperPath(path: String): String {
//        if (path.endsWith("/")) return path
//        return "$path/"
//    }
//
//    fun browseURL(url: String?) {
//        try {
//            val pb = ProcessBuilder()
//            pb.command("open", url)
//            pb.start() // .waitFor();
//        } catch (e: Exception) {
//            Logger.log(e)
//        }
//    }
//
    fun initShutdown() {
        try {
            val builder = ProcessBuilder()
            val lst = ArrayList<String>()
            lst.add("osascript")
            lst.add("-e")
            lst.add("tell app \"System Events\" to shut down")
            builder.command(lst)
            builder.start()
        } catch (e: Exception) {
            Logger.error(e)
        }
    }
}

